File: backend/src/main/java/pl/strava/analizator/domain/model/Activity.java
package pl.strava.analizator.domain.model;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import pl.strava.analizator.domain.vo.Lap;

@Getter
@Builder(toBuilder = true)
@AllArgsConstructor
public class Activity {

    private UUID id;
    private String externalId;
    private String source;
    private String sportType;
    private String name;
    private String description;

    private OffsetDateTime startedAt;
    private Integer elapsedTimeSec;
    private Integer movingTimeSec;

    private BigDecimal distanceM;
    private BigDecimal elevationGainM;
    private BigDecimal elevationLossM;

    private BigDecimal avgSpeedMs;
    private BigDecimal maxSpeedMs;

    private Short avgHeartrate;
    private Short maxHeartrate;

    private Short avgPowerW;
    private Short maxPowerW;

    private Short avgCadence;
    private Short maxCadence;

    private Integer calories;
    private BigDecimal avgTempC;

    private UUID gearId;
    private String summaryPolyline;
    private List<String> photoUrls;

    private List<String> tags;

    // Streams — time-series data for metric calculators
    private int[] powerStream;
    private int[] heartrateStream;
    private int[] cadenceStream;
    private double[] altitudeStream;
    private int[] timeStream;
    private double[] latStream;
    private double[] lngStream;
    private double[] distanceStream;
    private double[] velocityStream;

    private List<Lap> laps;

    private Instant createdAt;
    private Instant updatedAt;

    public boolean hasPowerData() {
        return powerStream != null && powerStream.length > 0;
    }

    public boolean hasHeartrateData() {
        return heartrateStream != null && heartrateStream.length > 0;
    }

    public boolean hasVelocityData() {
        return velocityStream != null && velocityStream.length > 0;
    }

    public boolean hasAltitudeData() {
        return altitudeStream != null && altitudeStream.length > 0;
    }

    public boolean hasGpsData() {
        return latStream != null && latStream.length > 0
                && lngStream != null && lngStream.length > 0;
    }
}

File: frontend/src/types/activity.ts
export interface ActivitySummary {
  id: string;
  externalId: string;
  sportType: string;
  name: string;
  startedAt: string;
  movingTimeSec: number;
  distanceM: number;
  elevationGainM: number | null;
  avgHeartrate: number | null;
  avgPowerW: number | null;
  avgSpeedMs: number | null;
  calories: number | null;
  summaryPolyline?: string | null;
  photoUrls?: string[] | null;
}

export interface ActivityDetail {
  id: string;
  externalId: string;
  source: string;
  sportType: string;
  name: string;
  description: string | null;
  startedAt: string;
  elapsedTimeSec: number;
  movingTimeSec: number;
  distanceM: number;
  elevationGainM: number | null;
  elevationLossM: number | null;
  avgSpeedMs: number | null;
  maxSpeedMs: number | null;
  avgHeartrate: number | null;
  maxHeartrate: number | null;
  avgPowerW: number | null;
  maxPowerW: number | null;
  avgCadence: number | null;
  maxCadence: number | null;
  calories: number | null;
  avgTempC: number | null;
  summaryPolyline: string | null;
  photoUrls?: string[] | null;
  powerStream: number[] | null;
  heartrateStream: number[] | null;
  cadenceStream: number[] | null;
  altitudeStream: number[] | null;
  timeStream: number[] | null;
  latStream: number[] | null;
  lngStream: number[] | null;
  distanceStream: number[] | null;
  velocityStream: number[] | null;
  laps: ActivityLap[] | null;
  metrics: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface ActivityLap {
  lapIndex: number;
  name?: string;
  startIndex: number | null;
  endIndex: number | null;
  distanceM: number;
  elapsedTimeSec: number;
  movingTimeSec: number;
  avgSpeedMs: number | null;
  maxSpeedMs: number | null;
  avgHeartrate: number | null;
  maxHeartrate: number | null;
  avgPowerW: number | null;
  maxPowerW: number | null;
  avgCadence: number | null;
  totalElevationGain: number | null;
  normalizedPowerW: number | null;
  variabilityIndex: number | null;
  powerDropPct: number | null;
  intensityClass: string | null;
}

export interface GeoJsonGeometry {
  type: string;
  coordinates: number[][];
}

export interface GeoJsonFeature {
  type: string;
  geometry?: GeoJsonGeometry;
  properties?: Record<string, string>;
  features?: GeoJsonFeature[];
}

export interface ActivityHeatmapBounds {
  south: number;
  west: number;
  north: number;
  east: number;
}

export interface HeatmapSegmentData {
  lat1: number;
  lon1: number;
  lat2: number;
  lon2: number;
  count: number;
}

export interface ActivityHeatmapData {
  segments: HeatmapSegmentData[];
  routeCount: number;
  bounds: ActivityHeatmapBounds | null;
  totalDistanceKm: number;
  maxCount: number;
  status: 'ready' | 'rebuilding';
}

export interface ActivitySummaryPage {
  items: ActivitySummary[];
  total: number;
  page: number;
  size: number;
  totalPages: number;
}

export interface ActivityTimelineEntry {
  year: number;
  month: number;
  count: number;
}
